flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java

```diff
@@ -423,6 +422,16 @@ private DataInputStatus emitNextNotReading(DataOutput<OUT> output) throws Except
         }
     }
 
+    private void initializeMainOutput(DataOutput<OUT> output) {
+        currentMainOutput = eventTimeLogic.createMainOutput(output, this::onWatermarkEmitted);
+        initializeLatencyMarkerEmitter(output);
+        lastInvokedOutput = output;
+        outputPendingSplits.forEach(
+                split -> currentMainOutput.createOutputForSplit(split.splitId()));
```

Comment:
I am wondering whether it might be problematic that you initialize all the outputs already but do not release them.

Previously the outputs were initialized one by one here [1] and released when moving to the next split[2]

[1] https://github.com/apache/flink/blob/106280e10a96d729943985986198b942446197d9/flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java#L327
[2] https://github.com/apache/flink/blob/106280e10a96d729943985986198b942446197d9/flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java#L195

Reply:
The output will be released only when the split is finished, so I think the logic is OK here, just advancing the output creation to the moment of split addition.

One concern in my mind is about adding an invalid splits to source reader. For example a split whose starting offset is ahead of ending offset, is added to the source reader, and the implementation of SplitReader just "swallows" this invalid split instead of marking it as finished, there will be an output leak.

Reply:
For your concern, we should update the javadoc of SplitReader to say that you always have to release.